feat: update "finding public content" (#104)
posit-dev/connect-cookbook: content/content-access-all/index.qmd

---
title: "Finding Public Content"

> execute:
>   eval: false

Reviewer: ```suggestion
```

Author: I initially left this in bc I thought it might be helpful when editing (i.e. don't execute even if the cell is "active"), but I realized… removing the dot to run cells, and saving, that's probably why my preview processes have stopped.

---

## Problem

> You need to identify all content on your Connect server with access permissions set to "Anyone - no login required". For each piece of public access content, you need to know the URL, content owner, and when it was last updated.

Reviewer: I think it will be enough to provide the code to "get" the content information. We can point them to the "Viewing Content Information" to understand the individual fields.

```suggestion
You need to find content that is public. Public content has no access controls and may be accessed by anyone with a link to the content on Connect.
```

Author: Yeah, I just put that in the recipe to justify why the recipe includes filtering. Like, it's perhaps useful in teaching to have the motivation in the example, and it also spells out more explicitly what the code below does. I was of two minds about phrasing it this way, but I looked at some of the examples in the other cookbooks, and their "Problems" do a similar thing where, they might be a bit more contrived but that also makes them more descriptive, e.g. "You have built a custom container object that internally holds a list, tuple, or some other iterable. You would like to make iteration work with your new container."

Author: Just to be clear, I'm gonna keep this as is but merge — if you still feel strongly we should change I'm happy to revisit :)

## Solution

> Use the Python or R SDK packages to get data on all content from Connect. Then, filter the data for public access content: the `access_type` column can take on three values: `all`, `logged_in`, and `acl`, and we only want `all`.

Reviewer: The bit on "Use the Python or R SDK packages" is redundant since that includes everything in the cookbook. I feel like the information on possible could also be removed since it isn't relevant to the task at hand. Maybe we should like directly to the API documentation for that information?

```suggestion
Find content where the *access type* is set to *all*.
```

Author: I agree that we don't need the line about the sdk packages. And we probably also don't need to spell out all the access types.

I'll accept the suggestion with a slight modification — using code spans instead of italics.

Side note I've long felt that our access type naming is confusingly inconsistent

Author: Hmm, reading this _in situ_ on the page, I think that a bit more text is sometimes better. Like, the whole point of the text is to add color and describe some of the context. When you remove the contextual information, it's like… why even include that?

So I removed all this text.

> You can also extend the recipe to create an HTML table suitable for use in a Shiny app or other dashboard.

Reviewer: Should this be in the "Extending" section?

Author: It could probably be. I had meant it to be, like, a reference / call ahead to that section, but the heading of that section is already descriptive.

Author: Removed this :

::: {.panel-tabset group="language"}

## Python

Use the Posit SDK to get all content from the server, filter for the access type you want, then prepare the data. The `owner` column contains a JSON object with multiple fields; we are only interested in the `username`. We also need to process the datetime field.

> ```{.python}

Reviewer: Should be...

```
from posit import connect

client = connect.Client()
```

And then to avoid using Pandas in this part of the example we can do:

`[content for content in client.content.find() if content.access_type == "all"]`

Feel free to break that up for readability.

Author: love a list comprehension

Author: Actually, it feels more appropriate to treat it as data and use the data frame filtering functionality… I did correct the import statement

Author: After trying it in situ, it actually felt more appropriate to a data science context to use the data frame filtering commands.
